### PR TITLE
[Fix] Skips adding a `$count` path if a similar `count()` function path exists

### PR DESCRIPTION
**Pull request by irvinesunday**
Fixes https://github.com/microsoft/OpenAPI.NET.OData/issues/347

### CODE REVIEW

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 1]
```diff
@@ -161,12 +186,48 @@ private void AppendPath(ODataPath path)
...
                     _allOperationPaths.Add(path);
                     break;
 
                 default:
                     return;
             }
+
+            bool DollarCountAndOperationPathsSimilar(ODataPath path1, ODataPath path2)
+            {
+                string countIdentifier = "count";
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:212]: move to a constant outside of the method

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 2]
```diff
@@ -161,12 +186,48 @@ private void AppendPath(ODataPath path)
...
+
+                if ((path1.Kind == ODataPathKind.DollarCount && 
+                    path2.Kind == ODataPathKind.Operation && path2.LastSegment.Identifier.Equals(countIdentifier, StringComparison.OrdinalIgnoreCase)) ||
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:215]:
```suggestion
                    path2.Kind == ODataPathKind.Operation && path2.LastSegment.Identifier.TrimEnd(parenthesis).EndsWith(countIdentifier, StringComparison.OrdinalIgnoreCase)) ||
```

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 3]
```diff
@@ -161,12 +186,48 @@ private void AppendPath(ODataPath path)
...
+                    (path2.Kind == ODataPathKind.DollarCount &&
+                    path1.Kind == ODataPathKind.Operation && path1.LastSegment.Identifier.Equals(countIdentifier, StringComparison.OrdinalIgnoreCase)))
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:217]:
```suggestion
                    path1.Kind == ODataPathKind.Operation && path1.LastSegment.Identifier.TrimEnd(parenthesis).EndsWith(countIdentifier, StringComparison.OrdinalIgnoreCase)))
```

and declare a static readonly char[] property that contains ( and )

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 4]
```diff
@@ -161,12 +186,48 @@ private void AppendPath(ODataPath path)
...
+                {
+                    return GetModifiedPathItemName(path1).Equals(GetModifiedPathItemName(path2), StringComparison.OrdinalIgnoreCase);
+                }
+
+                return false;                
+            }
+
+            string GetModifiedPathItemName(ODataPath path)
+            {
+                IEnumerable<ODataSegment> modifiedSegments = path.Take(path.Count - 1);
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:227]: return the original if ! any()

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 5]
```diff
@@ -161,12 +186,48 @@ private void AppendPath(ODataPath path)
 
                 case ODataPathKind.Operation:
                 case ODataPathKind.OperationImport:
+                    if (kind == ODataPathKind.Operation)
+                    {
+                        foreach (var kvp in _dollarCountPaths)
+                        {
+                            ODataPath dollarCountPath = kvp.Value.FirstOrDefault(p => DollarCountAndOperationPathsSimilar(p, path));
+                            if (dollarCountPath != null &&
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:194]:
```suggestion
                            if (kvp.Value.FirstOrDefault(p => DollarCountAndOperationPathsSimilar(p, path)) is ODataPath dollarCountPath &&
```

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 6]
```diff
@@ -142,6 +146,27 @@ private void AppendPath(ODataPath path)
                             nsList = new List<ODataPath>();
                             _allNavigationSourcePaths[navigationSourceSegment.EntityType] = nsList;
                         }
+                        
+                        if (kind == ODataPathKind.DollarCount)
+                        {                          
+                            ODataPath operationPath = _allOperationPaths.FirstOrDefault(p => DollarCountAndOperationPathsSimilar(p, path));
+                            
+                            if (operationPath != null)
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:154]:
```suggestion
                            if (_allOperationPaths.FirstOrDefault(p => DollarCountAndOperationPathsSimilar(p, path)) is ODataPath operationPath != null)
```

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for the quick fix here, appreciated. A couple of comments + a unit test and we should be good to go on this one.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 2 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:212]: Assuming the operation name here is "count", there are none and you can disregard. But if it's microsoft.graph.count() then it expands the tolerance of the match.
looking at the test cases you've updated, I'd say that change is not needed anymore. Go ahead and resolve if you agree.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 7]
```diff
@@ -142,6 +146,25 @@ private void AppendPath(ODataPath path)
                             nsList = new List<ODataPath>();
                             _allNavigationSourcePaths[navigationSourceSegment.EntityType] = nsList;
                         }
+                        
+                        if (kind == ODataPathKind.DollarCount)
+                        {                          
+                            if (_allOperationPaths.FirstOrDefault(p => DollarCountAndOperationPathsSimilar(p, path)) is ODataPath operationPath != null)
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:152]:
```suggestion
                            if (_allOperationPaths.FirstOrDefault(p => DollarCountAndOperationPathsSimilar(p, path)) is ODataPath operationPath)
```

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 3 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:214]: Following the comment above, does this still stand?

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 4 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:224]: The suggestion is unclear. Which original if ! any()?

Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:224]:
```suggestion
                if (!path.Any()) return path;
                IEnumerable<ODataSegment> modifiedSegments = path.Take(path.Count - 1);
```

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 3 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:214]: no if we assume the above is resolved with no changes, this one can be resolved too.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 4 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:224]: Went with `if (!path.Any()) return null;`

**Review by baywet — APPROVED** [review]
Thanks for taking in the changes!

File: src/Microsoft.OpenApi.OData.Reader/Common/Constants.cs [hunk 8]
```diff
@@ -189,5 +189,10 @@ internal static class Constants
         /// entity name
         /// </summary>
         public static string EntityName = "entity";
+
+        /// <summary>
+        /// count segment identifier
+        /// </summary>
+        public static string CountSegmentIdentifier = "count";
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Common/Constants.cs:196]:
```suggestion
        public const string CountSegmentIdentifier = "count";
```

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 9]
```diff
@@ -142,6 +146,25 @@ private void AppendPath(ODataPath path)
                             nsList = new List<ODataPath>();
                             _allNavigationSourcePaths[navigationSourceSegment.EntityType] = nsList;
                         }
+                        
+                        if (kind == ODataPathKind.DollarCount)
+                        {                          
+                            if (_allOperationPaths.FirstOrDefault(p => DollarCountAndOperationPathsSimilar(p, path)) is ODataPath operationPath)
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:152]:
```suggestion
                            if (_allOperationPaths.FirstOrDefault(p => DollarCountAndOperationPathsSimilar(p, path)) is ODataPath)
```

**Review by baywet — CHANGES_REQUESTED** [review]

**Review by baywet — DISMISSED** [review]

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 10]
```diff
@@ -142,6 +146,25 @@ private void AppendPath(ODataPath path)
                             nsList = new List<ODataPath>();
                             _allNavigationSourcePaths[navigationSourceSegment.EntityType] = nsList;
                         }
+                        
+                        if (kind == ODataPathKind.DollarCount)
+                        {                          
+                            if (_allOperationPaths.FirstOrDefault(p => DollarCountAndOperationPathsSimilar(p, path)) is ODataPath)
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:152]:
```suggestion
                            if (_allOperationPaths.FirstOrDefault(p => DollarCountAndOperationPathsSimilar(p, path)) is not null)
```

**Comment by baywet** [comment]
@irvinesunday a couple of unit tests are failing, can you look into that please?

**Review by baywet — APPROVED** [review]

**Pull request merged**